# ViralWasm-Consensus 

A client-side WebAssembly pipeline for viral consensus sequence calling. 

Visit: https://niema-lab.github.io/ViralWasm-Consensus/ to run the pipeline.

## ViralWasm-Consensus Offline
**Prerequisites**: Python (Can be downloaded here: https://www.python.org/downloads/)

To run the pipeline locally without internet, download the zip file containing the offline mode of ViralWasm-Consensus: https://github.com/Niema-Lab/ViralWasm-Consensus/raw/master/viralwasm-consensus-offline.zip. After, extract the folder, and right click on the `run_website.py` file and select "Open with Python" for Windows or "Open With -> Python Launcher" for Mac. 

Note: When starting the website for the first time offline, you may be prompted with a firewall warning. Click "Allow Access" to run the pipeline.

## Citing ViralWasm-Consensus

If you use ViralWasm-Consensus in your work, please cite:

> Ji D, Aboukhalil R, Moshiri N (2023). "ViralWasm: a client-side user-friendly web application suite for viral genomics." *Bioinformatics*. btae018. [doi:10.1093/bioinformatics/btae018](https://doi.org/10.1093/bioinformatics/btae018)

Please also cite the following tools used in ViralWasm-Epi:

### Biowasm

See https://biowasm.com/.

### ViralConsensus

> Moshiri N (2023). "ViralConsensus: A fast and memory-efficient tool for calling viral consensus genome sequences directly from read alignment data." *Bioinformatics*. btad317. [doi:10.1093/bioinformatics/btad317](https://doi.org/10.1093/bioinformatics/btad317)

### Minimap2

> Li H (2018). "Minimap2: pairwise alignment for nucleotide sequences." *Bioinformatics*. 34(18):3094–3100. [doi:10.1093/bioinformatics/bty191](https://doi.org/10.1093/bioinformatics/bty191)

### fastp

> Chen S, Zhou Y, Chen Y, Gu J (2018). "fastp: an ultra-fast all-in-one FASTQ preprocessor." *Bioinformatics*. 34(17):i884–i890. [doi:10.1093/bioinformatics/bty560](https://doi.org/10.1093/bioinformatics/bty560)